tests/nn/data_parallel/test_fsdp_fwd_multiple_times.py

```diff
@@ -0,0 +1,143 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+""" Test FSDP with forward pass with multiple calls of a module, like a loop. """
+
+import tempfile
+
+import pytest
+import torch
+import torch.multiprocessing as mp
+from torch.nn import Linear, Module, Sequential
+from torch.optim import Adam
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+from fairscale.nn.data_parallel.fully_sharded_data_parallel import TrainingState
+from fairscale.nn.misc import checkpoint_wrapper as ckpt
+from fairscale.nn.wrap import enable_wrap, wrap
+from fairscale.utils.testing import dist_init, skip_if_single_gpu, teardown, torch_version
+
+# To make it simpler to write and easier to understand, we do
+# FSDP and checkpointing in the Model classes explicitly.
+#
+# This means the models are created at the workers not by the
+# controller/main process.
+#
+# For this test, we don't do any numerical checking, so the weights
+# of the models at the each worker doesn't matter and we don't make
+# them the same.
+
+
+class ModelCkpt1Level(Module):
+    """ Model that has one level checkpoint."""
+
+    def __init__(self):
+        super().__init__()
+        self.child1 = ckpt(Linear(3, 3))
+        self.child2 = ckpt(Linear(3, 3))
+
+    def forward(self, x):
+        x = self.child1(x)
+        x = self.child2(x)
+        return x
+
+
+class ModelCkpt2Level(Module):
+    """ Model that has two level checkpoint."""
+
+    def __init__(self):
+        super().__init__()
+        self.child1 = ckpt(Sequential(ckpt(Linear(3, 3)), ckpt(Linear(3, 3))))
+        self.child2 = ckpt(Sequential(ckpt(Linear(3, 3)), ckpt(Linear(3, 3))))
+
+    def forward(self, x):
+        x = self.child1(x)
+        x = self.child2(x)
+        return x
+
+
+class ModelFSDP1Level(Module):
+    """ Model that has one level checkpoint."""
+
+    def __init__(self):
+        super().__init__()
+        self.child1 = wrap(Linear(3, 3))
+        self.child2 = wrap(Linear(3, 3))
+
+    def forward(self, x):
+        x = self.child1(x)
+        x = self.child2(x)
+        return x
+
+
+class ModelFSDPCkpt2Level(Module):
```

Comment:
very strange, this case should match the vissl case we saw yesterday, but it doesn't reproduce the error we saw. I think we need to beef up this test.